lib/src/test/scala/spinal/lib/ethernet/LoopbackCheck.scala

```diff
@@ -0,0 +1,214 @@
+      val dataQueue = new mutable.Queue[String]
+      def initPort(): Unit = {
+        dut.io.metaIn.payload.dstIpAddr #= 0
+        dut.io.metaIn.payload.dstPort #= 0
+        dut.io.metaIn.payload.dstMacAddr #= 0
+        dut.io.metaIn.payload.srcPort #= 0
+        dut.io.metaIn.payload.srcIpAddr #= 0
+        dut.io.metaIn.payload.srcMacAddr #= 0
+        dut.io.metaIn.dataLen #= 0
+        dut.io.metaIn.dataLen #= 0
+        dut.io.metaIn.valid #= false
+
+        dut.io.dataAxisOut.ready #= true
+        dut.io.metaOut.ready #= true
+
+        dut.io.dataAxisIn.valid #= false
+        dut.io.dataAxisIn.data #= 0
+        dut.io.dataAxisIn.last #= false
+        dut.io.dataAxisIn.keep #= 0
+        dut.io.dataAxisIn.user #= 0
+      }
+
+      initPort()
+      dut.clockDomain.waitRisingEdge(50)
+
+      println("init monitor")
+      def monitor(): Unit = {
+        while (true) {
+          if (dut.io.dataAxisOut.valid.toBoolean) {
+            val ref = dataQueue.dequeue()
+            val resData = dut.io.dataAxisOut.data.toBigInt.toString(16)
+            val resKeep = dut.io.dataAxisOut.keep.toLong
+            val resValid = (log10(resKeep + 1) / log10(2.0)).toInt
+            val resNum = if ((resData.length - resValid * 2) < 0) 0 else (resData.length - resValid * 2)
+            val res = resData.substring(resNum)
+            val refAppend = "0" * (64 - ref.length) + ref
+            val resAppend = "0" * (64 - res.length) + res
+            println(f"ref: $refAppend, res: $resAppend")
+            assert(refAppend == resAppend, s"require ${refAppend}, but return ${resAppend}")
+          }
+          dut.clockDomain.waitRisingEdge()
+        }
+      }
+
+      val enaMonitor = fork {
+        monitor()
+      }
+      println("start data drive")
+
+      for (idx <- 0 until simCfg.sendTimes) {
+        val sendDataBytes =
+          if (simCfg.useRandomPacket) Random.nextInt(1024).abs + 1
+          else simCfg.packetLen
+        println(idx + " times transaction...")
+        driveTransaction(sendDataBytes)
+      }
+
+      def driveTransaction(sendDataBytes: Int): Unit = {
+        val a = fork {
+          loadMeta(sendDataBytes)
+        }
+        val b = fork {
+          loadData(sendDataBytes)
+        }
+        a.join()
+        b.join()
+        dut.clockDomain.waitRisingEdge(Random.nextInt(20).abs)
+      }
+
+      def loadMeta(sendDataBytes: Int): Unit = {
+        dut.io.metaIn.payload.dstMacAddr #= simCfg.dstMacAddr.asHex
+        dut.io.metaIn.payload.dstIpAddr #= simCfg.dstIpAddr.asHex
+        dut.io.metaIn.payload.dstPort #= simCfg.dstPort.asHex
+        dut.io.metaIn.payload.srcMacAddr #= simCfg.srcMacAddr.asHex
+        dut.io.metaIn.payload.srcIpAddr #= simCfg.srcIpAddr.asHex
+        dut.io.metaIn.payload.srcPort #= simCfg.srcPort.asHex
+        dut.io.metaIn.payload.dataLen #= sendDataBytes
+        dut.io.metaIn.valid #= true
+        dut.clockDomain.waitRisingEdge()
```

Comment:
Prefer waitSampling to waitRisingEdge, as it take in acount the reset status and the clock edge kind.

Reply:
Got it : )